src/ui/public/visualize/loader.js

```diff
@@ -1,75 +1,170 @@
-      'time-range="timeRange" editor-mode="editorMode" show-spy-panel="showSpyPanel"></visualize>');
-    const visHtml = $compile(visEl)(scope);
-    container.html(visHtml);
+    const visHtml = $compile(visTemplate)(scope);
 
-    const handler = { destroy: scope.$destroy };
+    // If params specified cssClass, we will set this to the element.
+    if (params.cssClass) {
+      visHtml.addClass(params.cssClass);
+    }
 
-    return new Promise((resolve) => {
-      visEl.on('renderComplete', () => {
-        resolve(handler);
+    // Apply data- attributes to the element if specified
+    if (params.dataAttrs) {
+      Object.keys(params.dataAttrs).forEach(key => {
+        visHtml.attr(`data-${key}`, params.dataAttrs[key]);
       });
-    });
+    }
+
+    // If params.append was true append instead of replace content
+    if (params.append) {
+      container.append(visHtml);
+    } else {
+      container.html(visHtml);
+    }
 
+    return new EmbeddedVisualizeHandler(visHtml, scope);
   };
 
   return {
-    embedVisualizationWithId: async (el, savedVisualizationId, params) => {
-      return new Promise((resolve) => {
+    /**
+     * Renders a saved visualization specified by its id into a DOM element.
+     *
+     * @param {Element} element The DOM element to render the visualization into.
+     *    You can alternatively pass a jQuery element instead.
+     * @param {String} id The id of the saved visualization. This is the id of the
+     *    saved object that is stored in the .kibana index.
+     * @param {Object} params A list of parameters that will influence rendering.
+     *    See the `embedVisualizationWithSavedObject` documentation for a list of
+     *    all accepted parameters.
+     * @return {Promise.<EmbeddedVisualizeHandler>} A promise that resolves to the
+     *    handler for this visualization as soon as the saved object could be found.
+     */
+    embedVisualizationWithId: async (element, savedVisualizationId, params) => {
+      return new Promise((resolve, reject) => {
         savedVisualizations.get(savedVisualizationId).then(savedObj => {
-          renderVis(el, savedObj, params).then(handler => {
-            resolve(handler);
-          });
-        });
+          const handler = renderVis(element, savedObj, params);
+          resolve(handler);
+        }, reject);
       });
     },
+    /**
+     * Renders a saved visualization specified by its savedObject into a DOM element.
+     * In most of the cases you will need this method, since it allows you to specify
+     * filters, handlers, queries, etc. on the savedObject before rendering.
+     *
+     * @param {Element} element The DOM element to render the visualization into.
+     *    You can alternatively pass a jQuery element instead.
+     * @param {Object} savedObj The savedObject as it could be retrieved by the
+     *    `savedVisualizations` service.
+     * @param {Object} params A list of paramters that will influence rendering.
```

Comment:
Since we are rusing the shape of this options object, I'd consider using the typedef functionality to describe this shape instead. http://usejsdoc.org/tags-typedef.html

Then you can just reference it in your params, like

```
@typedef EmbedVisualizationParams
@property {AppState} appState
....


//and then reuse it in your method docs
@param {EmbedVisualizationParams} params

```

Reply:
Will change this, thanks.